src/BaGet.AWS/S3StorageService.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Amazon.S3;
+using Amazon.S3.Model;
+using BaGet.AWS.Configuration;
+using BaGet.Core.Entities;
+using BaGet.Core.Services;
+using Microsoft.Extensions.Options;
+using NuGet.Versioning;
+
+namespace BaGet.AWS
+{
+    public class S3StorageService : IStorageService
+    {
+        private readonly string _bucket;
+        private readonly string _prefix;
+        private readonly AmazonS3Client _client;
+        private const string Seperator = "/";
+
+        public S3StorageService(IOptions<S3StorageOptions> options, AmazonS3Client client)
+        {
+            _bucket = options.Value.Bucket;
+            _prefix = options.Value.Prefix;
+            _client = client;
+
+            if (!string.IsNullOrEmpty(_prefix) && !_prefix.EndsWith(Seperator))
+                _prefix += Seperator;
+        }
+
+        private string PrepareKey(string path)
+        {
+            return _prefix + path.Replace("\\", Seperator);
+        }
+
+        public async Task<Stream> GetAsync(string path, CancellationToken cancellationToken = default)
+        {
+            MemoryStream stream = new MemoryStream();
+
+            try
+            {
+                GetObjectResponse res = await _client.GetObjectAsync(_bucket, PrepareKey(path), cancellationToken);
+
+                await res.ResponseStream.CopyToAsync(stream);
+                stream.Seek(0, SeekOrigin.Begin);
+            }
+            catch (Exception)
+            {
+                stream.Dispose();
+
+                // TODO
+                throw;
+            }
+
+            return stream;
+        }
+
+        public Task<Uri> GetDownloadUriAsync(string path, CancellationToken cancellationToken = default)
+        {
+            string res = _client.GetPreSignedURL(new GetPreSignedUrlRequest
+            {
+                BucketName = _bucket,
+                Key = PrepareKey(path)
+            });
+
+            return Task.FromResult(new Uri(res));
+        }
+
+        public async Task<PutResult> PutAsync(string path, Stream content, string contentType, CancellationToken cancellationToken = default)
+        {
+            // TODO: Uploads should be idempotent. This should fail if and only if the blob
+            // already exists but has different content.
+
+            using (MemoryStream ms = new MemoryStream())
```

Comment:
Do you need to copy `content` to a memory stream? Could you instead pass the `content` stream directly to `PutObjectAsync`?

Reply:
We can't. Sometimes the `content` is a stream that doesn't support seeking, which is necessary for S3, since it requires the Length of the stream. There is a path through the S3 client code that doesn't require this, but I was unsuccesful in hitting it.

Essentially if you set the ContentLength header on the request object to something, it should use that length if acquiring it from `content` fails - but since we don't know the actual length in a DeflateStream, we can only set it to 0 or something very large. 0 doesn't quite work (uploaded data is actually 0 bytes) -- haven't tried setting it to something large.

Reply:
The only way I know of that approaches a "we don't know the length" is to buffer f.ex. ~25 MB at a time, and then do a Multipart Upload. But you'll still be buffering 25 MB at a time -- undoubtedly better than potentially gigabytes at a time.. :P

Reply:
AWS actually has a utility that does that - `TransferUtility`. Unfotunately, just like last I looked at it, it still doesn't support unknown lengths.

```
System.NotSupportedException: This operation is not supported.
   at System.IO.Compression.DeflateStream.get_Length()
   at Amazon.S3.Transfer.TransferUtilityUploadRequest.get_ContentLength() in E:\JenkinsWorkspaces\v3-trebuchet-release\AWSDotNetPublic\sdk\src\Services\S3\Custom\_bcl+coreclr+pcl\Transfer\TransferUtilityUploadRequest.cs:line 420
   at Amazon.S3.Transfer.TransferUtility.IsMultipartUpload(TransferUtilityUploadRequest request) in ...
```

Reply:
S3's issue tracker still shows no progress on this one - however, it _is_ possible as I mentioned, it's just a lot of code: fill a buffer till we hit limit, check if there is more data in the source (mark last-chunk), then send it, repeat..